### DISCUSSION TITLE
Using virtual hosts with RabbitMQ in Smallrye Reactive Messaging AMQP 1.0

### DISCUSSION THREAD

**Discussion by soframel** [smallrye/smallrye-reactive-messaging#2642] [Q&A]
Hello, 

Could you tell me if smallrye-reactive-messaging-amqp used in conjunction with RabbitMQ should work when working with queues inside virtual hosts?

Note: I know there is a specific version for rabbitmq in smallrye-reactive-messaging-rabbitmq, but I would like to check if the AMQP 1.0 version of the lib would fit our needs. In addition, since RabbitMQ 4 will support AMQP 1.0 by default, the change would probably have less impacts on our applications if we could start now the with the generic lib. 

I have tested with both properties `amqp-virtual-host` and `mp.messaging.incoming.<mychannel>.virtual-host` in a Quarkus application, like this: 

```
amqp-virtual-host=myvh
mp.messaging.incoming.myqueue.virtual-host=myvh
mp.messaging.incoming.myqueue.connector=smallrye-amqp
mp.messaging.incoming.myqueue.address=myqueue
mp.messaging.incoming.myqueue.durable=false
```

But messages in the queue "myqueue" inside virtual host "myvh" are never read by my application. And if I create a queue "myqueue" in the default vhost, messages are read correctly. 
In addition, my first tests for message producer / outgoing configs seem to show similar results (I get errors on RabbitMQ of the form "no exchange 'myqueue-exchange' in vhost '/' ", even though I am targetting another vhost). 

Am I missing something to configure virtual hosts?
(Note: I have found this issue in Quarkus: https://github.com/smallrye/smallrye-reactive-messaging/issues/474 but it's closed since a long time). 
Thank you.

**Comment by soframel** [comment]
Hello, 

a small update on this, since RabbitMQ 4.0 was released a few days ago and it supports AMQP 1.0 by default: 
After upgrading Smallrye reactive messaging to 4.24.0 (through Quarkus 3.14.4), I still have the same problem, at least for the producer (consumer not tested yet): with both properties specifying a virtual host, my connection is still refused by RabbitMQ 4 when connecting with AMQP 1.0. 

RabbitMQ log:
```
rabbit-1     | 2024-09-24 08:44:19.756430+00:00 [warning] <0.1209.0>                                             {symbol,<<"amqp:not-found">>},
rabbit-1     | 2024-09-24 08:44:19.756430+00:00 [warning] <0.1209.0>                                             {utf8,
rabbit-1     | 2024-09-24 08:44:19.756430+00:00 [warning] <0.1209.0>                                              <<"no exchange 'myqueue-exchange' in vhost '/'">>}
```

Even though my producer should try to connect to a specific vhost "myvhost" (where the exchange does exist). 

Producer properties: 
```
amqp-virtual-host=myvhost
...
mp.messaging.outgoing.myqueue-exchange.virtual-host=myvhost
mp.messaging.outgoing.myqueue-exchange.connector=smallrye-amqp
mp.messaging.outgoing.myqueue-exchange.address=/exchanges/myqueue-exchange
mp.messaging.outgoing.myqueue-exchange.use-anonymous-sender=false

```

is this a limitation or am I doing something wrong?

thank you.

**Reply by amusarra** [reply]
Hi @soframel 
Did you solve it? I opened a similar problem via issue #2954 where I also mentioned this discussion.

Thank you.

**Reply by soframel** [reply]
Hello, no, I did not, and we started using the smallrye-reactive-messaging-rabbitmq (in AMQP 0.9.1) instead (but I have not tested this since then)

**Reply by amusarra** [reply]
Hi @soframel 
You can resolve with this configuration:

`mp.messaging.outgoing.quote-requests.virtual-host=vhost:<your-virtual-host>`

and you can see on the RabbitMQ log this similar entry:

`2025-03-04 13:10:03.951210+00:00 [debug] <0.1636.0> AMQP 1.0 connection.open frame: hostname = vhost:amusarra-quarkus, extracted vhost = amusarra-quarkus, idle-time-out = undefined
`

Bye.

**Reply by soframel** [reply]
thanks, I'll test this!

**Reply by amusarra** [reply]
I have tried to configure the TLS but obtain an exception. You can see detail here #2960

**Reply by soframel** [reply]
Hello, I can confirm that for incoming channels, using the syntax "vhost:myvhost" now works (in smallrye reactive mesaging amqp 4.26.0 / Quarkus 3.18): 
mp.messaging.incoming.commands-channel.virtual-host=vhost:myvhost
